kafka-ui-api/src/main/java/com/provectus/kafka/ui/serdes/CustomSerdeLoader.java

```diff
@@ -0,0 +1,169 @@
+package com.provectus.kafka.ui.serdes;
+
+import com.provectus.kafka.ui.serde.api.PropertyResolver;
+import com.provectus.kafka.ui.serde.api.Serde;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import lombok.SneakyThrows;
+import lombok.Value;
+
+
+class CustomSerdeLoader {
+
+  @Value
+  static class CustomSerde {
+    Serde serde;
+    ClassLoader classLoader;
+  }
+
+  // serde location -> classloader
+  private final Map<Path, ClassLoader> classloaders = new ConcurrentHashMap<>();
+
+  @SneakyThrows
+  CustomSerde loadAndConfigure(String className,
+                               String location,
+                               PropertyResolver serdeProps,
+                               PropertyResolver clusterProps,
+                               PropertyResolver globalProps) {
+    Path locationPath = Path.of(location);
+    var serdeClassloader = createClassloader(locationPath);
+    var origCL = ClassloaderUtil.compareAndSwapLoaders(serdeClassloader);
+    try {
+      var serdeClass = serdeClassloader.loadClass(className);
+      var serde = (Serde) serdeClass.getDeclaredConstructor().newInstance();
+      serde.configure(serdeProps, clusterProps, globalProps);
+      return new CustomSerde(serde, serdeClassloader);
+    } finally {
+      ClassloaderUtil.compareAndSwapLoaders(origCL);
+    }
+  }
+
+  private static boolean isArchive(Path path) {
+    String archivePath = path.toString().toLowerCase(Locale.ROOT);
+    return archivePath.endsWith(".jar") || archivePath.endsWith(".zip");
+  }
+
+  @SneakyThrows
+  private static List<URL> findArchiveFiles(Path location) {
+    List<URL> archives = new ArrayList<>();
+    if (isArchive(location)) {
+      archives.add(location.toUri().toURL());
+    } else if (Files.isDirectory(location)) {
+      try (var files = Files.walk(location)) {
+        var paths = files.filter(CustomSerdeLoader::isArchive).collect(Collectors.toList());
+        for (Path path : paths) {
```

Comment:
let's replace for with stream's foreach?

Reply:
it trows checked exception, so code will be ugly